src/stores/RoomViewStore.js

```diff
@@ -128,16 +137,36 @@ class RoomViewStore extends Store {
         this._setState({
             joining: true,
         });
-        MatrixClientPeg.get().joinRoom(this._state.roomId, payload.opts).then(
-        () => {
-            this._setState({
-                joining: false,
+        MatrixClientPeg.get().joinRoom(this._state.roomId, payload.opts).done(() => {
+            dis.dispatch({
+                action: 'joined_room',
+                room_id: this._state.roomId,
             });
         }, (err) => {
-            this._setState({
-                joining: false,
-                joinError: err,
+            dis.dispatch({
+                action: 'join_room_error',
+                room_id: this._state.roomId,
+                err: err,
             });
+            const msg = err.message ? err.message : JSON.stringify(err);
+            const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
+            Modal.createDialog(ErrorDialog, {
```

Comment:
It feels quite wrong to create modals in the store. I would at least move this to `_onRoomViewStoreUpdate` of `RoomView`, but I am surprised that we weren't already doing this somewhere else...